File: app/models/timeline.py
"""
Timeline models for issue feed, interactions, and analytics.
"""

from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional, List, Dict
from enum import Enum


class VoteType(str, Enum):
    """Vote types."""
    UPVOTE = "UPVOTE"
    DOWNVOTE = "DOWNVOTE"


class SourceType(str, Enum):
    """Source types for reports."""
    CITIZEN = "CITIZEN"
    WEBSITE_SCRAPER = "WEBSITE_SCRAPER"
    API = "API"
    ADMIN = "ADMIN"


class TimelineIssue(BaseModel):
    """Issue model for timeline feed."""
    id: str
    title: str
    description: str
    issue_type: str
    severity: str
    confidence: str
    status: str
    city: Optional[str] = None
    locality: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    created_at: str
    updated_at: str
    
    # Scores
    popularity_score: int = Field(default=0, description="Popularity score (upvotes - downvotes)")
    confidence_score: float = Field(default=0.0, ge=0.0, le=1.0, description="AI confidence score")
    priority_score: Optional[int] = Field(None, ge=0, le=100, description="Priority score")
    
    # Interaction counts
    upvote_count: int = Field(default=0)
    downvote_count: int = Field(default=0)
    comment_count: int = Field(default=0)
    report_count: int = Field(default=1)
    
    # User interaction (if authenticated)
    user_vote: Optional[VoteType] = None
    is_bookmarked: bool = Field(default=False)
    
    # Source information
    sources: List[Dict] = Field(default_factory=list, description="List of sources (citizens, scrapers, etc.)")
    
    # Media
    media_urls: List[str] = Field(default_factory=list)


class CommentCreate(BaseModel):
    """Model for creating a comment."""
    issue_id: str
    text: str = Field(..., min_length=1, max_length=1000)
    parent_comment_id: Optional[str] = None  # For nested comments


class CommentResponse(BaseModel):
    """Comment response model."""
    id: str
    issue_id: str
    user_id: Optional[str] = None
    user_phone: Optional[str] = None
    text: str
    parent_comment_id: Optional[str] = None
    created_at: str
    upvote_count: int = Field(default=0)
    downvote_count: int = Field(default=0)
    user_vote: Optional[VoteType] = None


class VoteRequest(BaseModel):
    """Vote request model."""
    issue_id: str
    vote_type: VoteType


class IssueAnalytics(BaseModel):
    """Analytics data for an issue."""
    issue_id: str
    
    # Scores
    popularity_score: int
    confidence_score: float
    priority_score: Optional[int]
    
    # Source breakdown
    source_breakdown: Dict[str, int] = Field(default_factory=dict)  # {CITIZEN: 5, WEBSITE_SCRAPER: 2}
    
    # Time series data for charts
    reports_over_time: List[Dict] = Field(default_factory=list)  # [{date: "2024-01-15", count: 5}]
    confidence_over_time: List[Dict] = Field(default_factory=list)  # [{date: "2024-01-15", confidence: 0.8}]
    votes_over_time: List[Dict] = Field(default_factory=list)  # [{date: "2024-01-15", upvotes: 10, downvotes: 2}]
    
    # Distribution data for pie charts
    issue_type_distribution: Dict[str, int] = Field(default_factory=dict)
    severity_distribution: Dict[str, int] = Field(default_factory=dict)
    status_distribution: Dict[str, int] = Field(default_factory=dict)
    
    # Heatmap data (location-based)
    location_heatmap: List[Dict] = Field(default_factory=list)  # [{lat: 18.5, lng: 73.8, intensity: 5}]
    
    # AI metadata
    ai_metadata: Optional[Dict] = None
    
    # Comments
    comments: List[CommentResponse] = Field(default_factory=list)
    
    # Total interactions
    total_upvotes: int = Field(default=0)
    total_downvotes: int = Field(default=0)
    total_comments: int = Field(default=0)
    total_reports: int = Field(default=0)

File: app/lib/timeline.ts
/**
 * Timeline API functions.
 */

import { API_BASE_URL } from "./config";
import { getAuthToken, getCurrentUser } from "./auth";

export interface TimelineIssue {
  id: string;
  title: string;
  description: string;
  issue_type: string;
  severity: string;
  confidence: string;
  status: string;
  city?: string;
  locality?: string;
  latitude?: number;
  longitude?: number;
  created_at: string;
  updated_at: string;
  popularity_score: number;
  confidence_score: number;
  priority_score?: number;
  upvote_count: number;
  downvote_count: number;
  comment_count: number;
  report_count: number;
  user_vote?: "UPVOTE" | "DOWNVOTE";
  is_bookmarked: boolean;
  sources: Array<{ type: string; count: number; description: string }>;
  media_urls: string[];
}

export interface IssueAnalytics {
  issue_id: string;
  popularity_score: number;
  confidence_score: number;
  priority_score?: number;
  source_breakdown: Record<string, number>;
  reports_over_time: Array<{ date: string; count: number }>;
  confidence_over_time: Array<{ date: string; confidence: number }>;
  votes_over_time: Array<{ date: string; upvotes: number; downvotes: number }>;
  issue_type_distribution: Record<string, number>;
  severity_distribution: Record<string, number>;
  status_distribution: Record<string, number>;
  location_heatmap: Array<{ lat: number; lng: number; intensity: number }>;
  ai_metadata?: Record<string, any>;
  comments: Comment[];
  total_upvotes: number;
  total_downvotes: number;
  total_comments: number;
  total_reports: number;
}

export interface Comment {
  id: string;
  issue_id: string;
  user_id?: string;
  user_phone?: string;
  text: string;
  parent_comment_id?: string;
  created_at: string;
  upvote_count: number;
  downvote_count: number;
  user_vote?: "UPVOTE" | "DOWNVOTE";
}

async function fetchWithAuth(url: string, options: RequestInit = {}) {
  const token = getAuthToken();
  const user = getCurrentUser();
  
  const headers: Record<string, string> = {
    "Content-Type": "application/json",
    ...(options.headers as Record<string, string>),
  };
  
  if (user?.id) {
    headers["X-User-ID"] = user.id;
  }
  
  const response = await fetch(url, {
    ...options,
    headers,
  });
  
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: "Unknown error" }));
    throw new Error(error.detail || `API error ${response.status}`);
  }
  
  return response.json();
}

export async function fetchTimelineFeed(city?: string, limit = 50): Promise<TimelineIssue[]> {
  const params = new URLSearchParams();
  if (city) params.append("city", city);
  params.append("limit", limit.toString());
  
  const url = `${API_BASE_URL}/timeline/feed?${params.toString()}`;
  return fetchWithAuth(url);
}

export async function fetchIssueAnalytics(issueId: string): Promise<IssueAnalytics> {
  const url = `${API_BASE_URL}/timeline/issue/${issueId}/analytics`;
  return fetchWithAuth(url);
}

export async function voteOnIssue(issueId: string, voteType: "UPVOTE" | "DOWNVOTE"): Promise<{
  success: boolean;
  action: string;
  upvote_count: number;
  downvote_count: number;
  popularity_score: number;
  user_vote?: string;
}> {
  const url = `${API_BASE_URL}/timeline/issue/${issueId}/vote?vote_type=${voteType}`;
  return fetchWithAuth(url, { method: "POST" });
}

export async function addComment(issueId: string, text: string, parentCommentId?: string): Promise<Comment> {
  const url = `${API_BASE_URL}/timeline/comment`;
  return fetchWithAuth(url, {
    method: "POST",
    body: JSON.stringify({
      issue_id: issueId,
      text,
      parent_comment_id: parentCommentId,
    }),
  });
}

export async function fetchIssueComments(issueId: string): Promise<Comment[]> {
  const url = `${API_BASE_URL}/timeline/issue/${issueId}/comments`;
  return fetchWithAuth(url);
}
